File: frontend/src/services/aportes.ts
import { http } from "./http"

export interface Aporte {
  id: number
  ano_id: number
  mes: number
  data: string
  ativo_id: number
  tipo_operacao: "compra" | "venda"
  quantidade: number
  preco_unitario: number
  taxas: number
  moeda: "BRL" | "USD"
  cotacao_usd_brl: number | null
  valor_total: number
  valor_total_brl: number
  conta_id: number | null  // 🆕 agora opcional
  descricao: string | null
  criado_em: string
  atualizado_em: string
}

export interface ResumoMensalAportes {
  mes: number
  total_brl: number
  qtd_operacoes: number
  qtd_compras: number
  qtd_vendas: number
  total_compras_brl: number
  total_vendas_brl: number
}

export interface ResumoPorAtivo {
  ativo_id: number
  ticker: string
  nome: string
  qtd_operacoes: number
  total_brl: number
  quantidade_acumulada: number | null
}

export interface FiltroAportes {
  ano_id?: number
  mes?: number
  ativo_id?: number
  tipo_operacao?: string
}

export const aportesService = {
  listar: (f: FiltroAportes = {}) =>
    http.get<Aporte[]>("/aportes", { params: f }).then(r => r.data),
  obter: (id: number) =>
    http.get<Aporte>(`/aportes/${id}`).then(r => r.data),
  criar: (p: Partial<Aporte>) =>
    http.post<Aporte>("/aportes", p).then(r => r.data),
  atualizar: (id: number, p: Partial<Aporte>) =>
    http.put<Aporte>(`/aportes/${id}`, p).then(r => r.data),
  excluir: (id: number) => http.delete(`/aportes/${id}`),
  resumoMensal: (ano_id: number) =>
    http.get<ResumoMensalAportes[]>("/aportes/resumo-mensal",
      { params: { ano_id } }).then(r => r.data),
  porAtivo: (ano_id: number, tipo_operacao?: string) =>
    http.get<ResumoPorAtivo[]>("/aportes/por-ativo",
      { params: { ano_id, tipo_operacao } }).then(r => r.data),
}

File: backend/app/schemas/aportes.py
from datetime import date
from pydantic import Field
from app.schemas.common import ORMBase, TimestampSchema


class AporteBase(ORMBase):
    ano_id: int
    mes: int = Field(ge=1, le=12)
    data: date
    ativo_id: int
    tipo_operacao: str = Field(pattern="^(compra|venda)$")
    quantidade: float = Field(gt=0)
    preco_unitario: float = Field(gt=0)
    taxas: float = Field(default=0, ge=0)
    moeda: str = Field(pattern="^(BRL|USD)$")
    # 🆕 Conta opcional
    conta_id: int | None = None
    descricao: str | None = None
    cotacao_usd_brl: float | None = Field(default=None, gt=0)


class AporteCreate(AporteBase):
    pass


class AporteUpdate(ORMBase):
    ano_id: int | None = None
    mes: int | None = Field(default=None, ge=1, le=12)
    data: date | None = None
    ativo_id: int | None = None
    tipo_operacao: str | None = None
    quantidade: float | None = Field(default=None, gt=0)
    preco_unitario: float | None = Field(default=None, gt=0)
    taxas: float | None = Field(default=None, ge=0)
    moeda: str | None = None
    cotacao_usd_brl: float | None = None
    conta_id: int | None = None
    descricao: str | None = None


class AporteOut(AporteBase, TimestampSchema):
    id: int
    valor_total: float
    valor_total_brl: float